Narrow constraints to `IsRecentEra era` in `Shelley.Transaction` module

Anviking opened the pull request:
- Replace `IsShelleyBasedEra` and `IsCardanoEra` and similar with `IsRecentEra` / `RecentEra era`
- [x] Error appropriately from signTransaction with non-recent tx formats

### Comments

<!-- Additional comments, links, or screenshots to attach, if any. -->

### Issue Number

ADP-2353 — increase consistency first

Anviking committed:
Narrow constraints to `IsRecentEra era` in `Shelley.Transaction` module

Anviking commented on lib/wallet/src/Cardano/Wallet/Shelley/Transaction.hs:
```diff
@@ -600,32 +573,15 @@ newTransactionLayer keyF networkId = TransactionLayer
+                                policyResolver policyKeyM stakingScriptResolver
+                                addressResolver inputResolver (body, wits)
+                            & sealedTxFromCardano'
+                        Nothing -> sealedTx
```
Should probably error... It matters more now that we no longer support signing  `Shelley, ..., Alonzo` txs

jonathanknowles replied:
See: 8f70f9bf73ae5c8beb3261ea917818130ea8ba09

jonathanknowles committed:
Fix formatting in `mkTransaction`.

- 80-column line length limit.
- 4-space indents.

jonathanknowles committed:
Fix documentation comments for `mkTransaction`.

- Use haddock syntax.
- Use correct grammar and punctuation.

jonathanknowles commented on lib/balance-tx/lib/internal/Internal/Cardano/Write/Tx.hs:
```diff
@@ -357,7 +357,9 @@ type RecentEraLedgerConstraints era =
-    -> ((RecentEraLedgerConstraints (CardanoApi.ShelleyLedgerEra era)) => a)
+    -> ( ( IsRecentEra era
+         , (RecentEraLedgerConstraints (CardanoApi.ShelleyLedgerEra era))
+         ) => a)
```
At first glance this seems a bit awkward to read (and to format).

Why not define a separate constraint synonym:
```hs
type RecentEraConstraints era =
    ( IsRecentEra era
    , RecentEraLedgerConstraints (CardanoApi.ShelleyLedgerEra era)
    )
```

Then we can write:
```hs
withConstraints :: RecentEra era -> (RecentEraConstraints era => a) -> a
```

jonathanknowles committed:
Fix documentation comment within `buildAndSignTransaction`.

- Revise fluency, grammar, and punctuation.
- Re-flow to fit within the line length limit.

jonathanknowles committed:
Fix formatting in `mkUnsignedTransaction`.

- 80-column line length limit.
- 4-space indents.

jonathanknowles reviewed (commented):
Hi @Anviking 

Would love to discuss this during our call later, so I can better understand the motivation and direction!

Nevertheless, I've put some suggestions in https://github.com/cardano-foundation/cardano-wallet/pull/4215. These suggestions are mainly fixes to formatting and documentation.

jonathanknowles committed:
Throw error in event that `addVkWitnesses` is called for old era.

Anviking committed:
Introduce `RecentEraConstraints`
